### DISCUSSION TITLE
How to return js expressions from a macro?

### DISCUSSION THREAD

**Discussion by ranfdev** [squint-cljs/cherry#131] [Q&A] [answered]
The code 
```clojure
(prn #js [1 2])
```
correctly compiles to
```js
cherry_core.prn.call(null, [1, 2]);
```

But now let's say that I want to refactor said code with a macro, by doing
```clojure
;; file macros.cljs
(ns macros)
(defmacro append-two-then-as-js [x]
    (quote #js [~x 2])
    
;; file main.js 
(ns main
  (:require-macros
   [macros :refer [append-two-then-as-js]]))
(prn (append-two-then-as-js 1))
```

The resulting code is
```js
cherry_core.prn.call(null, [object Object]);
```

**Problem:**
The `#js` is a reader macro that gets expanded during the evaluation of the macro `append-two-then-as-js`. But that is arguably useless, because now the compiled code contains `[object Object]`.

**Question:**
How can we write macros returning `#js` expressions? I can't find a way to write a macro returning `#js [1 2]` as is

**Comment by borkdude** [comment]
You can't do that. Just don't use `#js [1 2]` in a macro, since the reader is executed at macro expansion time. If you want to return code that constructs a JS array at runtime, just write something like: 

```
`(array ~x 2)
```

**Reply by ranfdev** [reply]
Thanks, I didn't know about that possibility.

Do you think it would be possible to write a generic `(as-js form)` macro, capable of doing the same thing as #js, by manually checking the type of `form` and calling the needed function, as `(array x)` for vectors...?

Would it make sense to have this theoretical `(as-js form)` macro, and implement `#js` in terms of it, so that `#js [1 2]` gets translated to `(as-js [1 2])`?

**Reply by borkdude** [reply]
The thing you're looking for is probably `clj->js`

**Reply by ranfdev** [reply]
Mmm... Not really, because I can't call `clj->js` for every object or array literal, for performance reasons. But I will use a macro to call `(array form)`, `(js-obj form)` and `(clj->js)` as needed. 

Thanks!

**Comment by ranfdev** [comment]
`(js-obj)` and `(array)` though are translated in runtime function calls, while `#js {}` and `#js []` don't.
I really can't find a way to return js literals from a macro

**Reply by borkdude** [reply, marked as answer]
Actually after digging in some more, if you would return from the macro:

```
(vary-meta {} assoc :cherry.compiler/js true)
```

on the returned map, I think cherry might already translate it to a JS object at compile time.

**Reply by ranfdev** [reply]
Indeed, that works, and that's exactly what I was looking for. Thanks

**Comment by borkdude** [comment]
@ranfdev In the latest cherry version the key changed to `:squint.compiler-common/js`. But instead I think it's better to just optimize `js-obj` and `array` to emit literals when possible. I don't think CLJS has a way to emit JS objects from macros either other than `js-obj` and `array`.

**Reply by borkdude** [reply]
Here's the issue: https://github.com/squint-cljs/cherry/issues/141

**Reply by ranfdev** [reply]
Thanks I'll look into it
